src/lib.rs

```diff
@@ -806,3 +934,241 @@ fn simple_build() {
+    data_offset: u64,
+
+    /// data buffer size
+    data_size: u64,
+
+    aux_head: u64,
+    aux_tail: u64,
+    aux_offset: u64,
+    aux_size: u64,
+}
+
+/// A stream of samples being sent to us from the kernel. These samples represent instantaneous
+/// states or events concerning the process(es) being profiled.
+///
+/// Internally the samples are queued up in a ring-buffer. The kernel writes samples into the
+/// buffer, and [`SampleStream.read`] deques them. If the buffer it full, the kernel will overwrite
+/// old samples effectively dropping them.
+pub struct SampleStream {
+    file: File,
+    mapped_memory: *mut c_void,
+    attrs: sys::bindings::perf_event_attr,
+}
+
+unsafe impl Send for SampleStream {}
+unsafe impl Sync for SampleStream {}
+
+impl SampleStream {
+    fn new(attrs: sys::bindings::perf_event_attr, file: File) -> std::io::Result<Self> {
+        let mapped_memory = check_syscall(|| unsafe {
+            mmap(
+                std::ptr::null_mut(),
+                SAMPLE_BUFFER_SIZE,
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED,
+                file.as_raw_fd(),
+                0,
+            ) as isize
+        })? as *mut c_void;
+        Ok(Self {
+            file,
+            mapped_memory,
+            attrs,
+        })
+    }
+
+    /// Begin sampling. If read is called before the stream is enabled, it will block until it is.
+    pub fn enable(&self) -> io::Result<()> {
+        check_syscall(|| unsafe { sys::ioctls::ENABLE(self.file.as_raw_fd(), 0) }).map(|_| ())
+    }
+
+    // If a thread is asleep in read, calling this function does not wake it up. As such the only
+    // utility to this function is to tell the kernel to stop sending us events when we destroy the
+    // stream.
+    fn disable(&self) -> io::Result<()> {
+        check_syscall(|| unsafe { sys::ioctls::DISABLE(self.file.as_raw_fd(), 0) }).map(|_| ())
+    }
+
+    /// Pop a sample from the buffer. If the buffer is empty, blocking waiting for there to be one
+    /// to return. If a timeout is given, it only blocks for up to the given timeout. When the
+    /// timeout it reached, None is returned.
+    pub fn read(&self, timeout: Option<std::time::Duration>) -> io::Result<Option<PerfRecord>> {
+        // XXX There is definitely a way to implement this function that has less copies and no
+        // heap allocations. If we made some circular reader type we could decode records directly
+        // from that. It just makes things a bit trickier.
+
+        // wait for there to be data in the buffer, or the timeout.
+        if wait_for_readable_or_timeout(&self.file, timeout) {
+            return Ok(None);
+        }
+
+        // The kernel gives us records in a ring buffer. As the kernel adds records to the head, we
+        // are consuming from the tail. If the buffer is full, the kernel drops records.
+        let header: *mut PerfEventMmapPage = unsafe { std::mem::transmute(self.mapped_memory) };
```

Comment:
This is just turning a `*mut c_void` into a `*mut PerfEventMmapPage`, right? Oddly, you can just use `as *mut PerfEventMmapPage` to do this, no `transmute` required, and hence no `unsafe` required.

Reply:
fixed